boot/flags.go

```diff
@@ -0,0 +1,251 @@
+			Role: bootloader.RoleRecovery,
+		}
+		bl, err := bootloader.Find(InitramfsUbuntuSeedDir, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		m, err := bl.GetBootVars("snapd_boot_flags")
+		if err != nil {
+			return nil, err
+		}
+
+		return splitFlagString(m["snapd_boot_flags"]), nil
+
+	default:
+		return nil, fmt.Errorf("internal error: unsupported mode %q", mode)
+	}
+}
+
+// InitramfsSetBootFlags sets the boot flags for the current boot in the /run
+// file that will be consulted in userspace by BootFlags() below. It is meant to
+// be used only from the initramfs.
+func InitramfsSetBootFlags(flags []string) error {
+	// when we are processing boot flags from the initramfs, don't enforce the
+	// allow list such that an old initramfs doesn't drop new boot flags that
+	// userspace snapd understands
+
+	// we do however drop empty boot flags to protect against the fact that
+	// in install mode, the bootenv is untrusted input, so we want some
+	// assurance that garbage is not being propagated through the system
+	// inadvertently
+	nonEmptyFlags := dropEmptyBootFlags(flags)
+	s := serializeFlagString(nonEmptyFlags)
+
+	if err := os.MkdirAll(dirs.SnapBootstrapRunDir, 0755); err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(snapBootstrapBootFlagsFile, []byte(s), 0644)
+}
+
+// BootFlags returns the current set of boot flags that were active when this
+// device was booted, using a file in /run/ instead of the source of truth. This
+// is to reduce ambiguity about which flags are active for this boot versus the
+// next boot.
+// Only to be used on UC20+ systems with recovery systems.
+func BootFlags(dev Device) ([]string, error) {
+	if !dev.HasModeenv() {
+		return nil, errNotUC20
+	}
+
+	// read the file that the initramfs wrote in /run
+	b, err := ioutil.ReadFile(snapBootstrapBootFlagsFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return splitFlagString(string(b)), nil
+}
+
+// NextBootFlags returns the set of boot flags that are applicable for the next
+// boot. This information always comes from the modeenv, since the only
+// situation where boot flags are set for the next boot and we query their state
+// is during run mode. The next boot flags for install mode are not queried
+// during prepare-image time, since they are only written to the bootenv at
+// prepare-image time.
+// Only to be used on UC20+ systems with recovery systems.
+// TODO: should this accept a modeenv that was previously read from i.e.
+// devicestate manager?
+func NextBootFlags(dev Device) ([]string, error) {
+	if !dev.HasModeenv() {
+		return nil, errNotUC20
+	}
+
+	m, err := ReadModeenv("")
+	if err != nil {
+		return nil, err
+	}
+
+	return m.BootFlags, nil
+}
+
+// SetNextBootFlags sets the boot flags for the next boot to take effect after
+// rebooting. This information always gets saved to the modeenv.
+// Only to be used on UC20+ systems with recovery systems.
+func SetNextBootFlags(dev Device, rootDir string, flags []string) error {
```

Comment:
we don't need this one yet, do we? I wonder if it should really be a set of options to the Ensure functions instead? It just that setting the flags make sense only for some modes but here we are not setting the mode, not we can check it because it will likely be the previous one?

Reply:
yes, we don't need this immediately, but we will want this eventually I think, and having this be it's own function here makes easier testing. I can unexport it and the Ensure*() functions can use it internally when needed.